### PR TITLE
[WIP] Interpret tokio

### LINKED ISSUE #602: can't call foreign function: sched_getaffinity
```rust
fn main() {
    tokio::run(futures::future::ok::<_, ()>(()))
}
```

```
error[E0080]: constant evaluation error: can't call foreign function: sched_getaffinity
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/num_cpus-1.9.0/src/lib.rs:335:17
    |
335 |     if unsafe { libc::sched_getaffinity(0, std::mem::size_of::<libc::cpu_set_t>(), &mut set) } == 0 {
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function: sched_getaffinity
    |
    = note: inside call to `num_cpus::get_num_cpus` at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/num_cpus-1.9.0/src/lib.rs:63:5
    = note: inside call to `num_cpus::get` at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.14/src/runtime/threadpool/builder.rs:72:28
    = note: inside call to `tokio::runtime::threadpool::builder::Builder::new` at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.14/src/runtime/threadpool/mod.rs:142:9
    = note: inside call to `tokio::runtime::threadpool::Runtime::new` at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.14/src/runtime/threadpool/mod.rs:103:23
note: inside call to `tokio::runtime::threadpool::run::<futures::future::result_::FutureResult<(), ()>>` at src/main.rs:2:5
   --> src/main.rs:2:5
    |
2   |     tokio::run(futures::future::ok::<_, ()>(()))
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: inside call to `main` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:64:34
    = note: inside call to closure at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:53
    = note: inside call to closure at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:297:40
    = note: inside call to `std::panicking::try::do_call::<[closure@DefId(1/1:1899 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:293:5
    = note: inside call to `std::panicking::try::<i32, [closure@DefId(1/1:1899 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe]>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:388:9
    = note: inside call to `std::panic::catch_unwind::<[closure@DefId(1/1:1899 ~ std[82ff]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:25
    = note: inside call to `std::rt::lang_start_internal` at /root/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:64:5
    = note: inside call to `std::rt::lang_start::<()>`

```

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #602

### CODE REVIEW

File: src/fn_call.rs [hunk 1]
```diff
@@ -15,7 +15,23 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a+'mir>: crate::MiriEvalContextExt<'a,
         dest: Option<PlaceTy<'tcx, Borrow>>,
         ret: Option<mir::BasicBlock>,
     ) -> EvalResult<'tcx, Option<&'mir mir::Mir<'tcx>>> {
+
+        // HACK: normal evaluation never sees `ReservedForIncrCompCache`, so we use it
+        // to signify a dlsym loaded thing
```
Comment (RalfJung) [inline src/fn_call.rs]: OMG WTF.^^

Okay so at the least could we wrap this awful hack into two functions `dlsym_to_instance` and `instance_to_dlsym` (or so)?  Currently the other half of this hack is 200 lines away in the same file without any comment.

The better way, of course, would be to change `create_fn_alloc` such that it takes something like
```rust
enum MiriFunction {
  Instance(Instance),
  DlSym(String),
}
```
Then we'd need no hack.

File: src/fn_call.rs [hunk 2]
```diff
@@ -693,4 +729,23 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a+'mir>: crate::MiriEvalContextExt<'a,
...
+        &mut self,
+        id: u32,
+        args: &[OpTy<'tcx, Borrow>],
+        dest: PlaceTy<'tcx, Borrow>,
+        ret: mir::BasicBlock,
+    ) -> EvalResult<'tcx> {
+        let this = self.eval_context_mut();
+        match id {
+            // epoll_create1
+            0 => {
```
Comment (RalfJung) [inline src/fn_call.rs]: No magic constants please.  Seems like a C-like enum would be appropriate here, or so?

File: src/fn_call.rs (hunk 1 above)
Comment (oli-obk) [reply src/fn_call.rs]: I know... but the compiler and const eval have zero profit from supporting this but would still have to live with the complexity.

File: src/operator.rs [hunk 3]
```diff
@@ -157,10 +157,12 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for super::MiriEvalContext<'a, 'mir, '
...
-                // Case I: Comparing with NULL
-                if bits == 0 {
-                    // Test if the ptr is in-bounds. Then it cannot be NULL.
-                    // Even dangling pointers cannot be NULL.
+                // Case I: Comparing with small integers
+                if bits <= 1 {
+                    // Test if the ptr is in-bounds. Then it cannot be a small integer.
+                    // Even dangling pointers cannot be small integers.
+                    // The only exception is wasm and some embedded devices, but that's a different
+                    // story: https://github.com/rust-lang/rust/issues/57897
```
Comment (RalfJung) [inline src/operator.rs]: I don't like this.  Enabling such platform-specific assumptions should be somehow, at the very least, tied to the platform -- as in, depending on the target we should do this or not.  I think we should also have a flag ("strict mode" or so) to not make such assumptions.

For platforms where we make this assumption, probably it applies at least to `bits < 4096`, doesn't it?

File: tokio_test/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "tokio_test"
+version = "0.1.0"
+authors = ["Oliver Scherer <github35764891676564198441@oli-obk.de>"]
+edition = "2018"
+
+[dependencies]
+tokio = "0.1.14"
+futures = "0.1.25"
```
Comment (RalfJung) [inline tokio_test/Cargo.toml]: Can we have these in a subdir or so?  We are getting too many root dirs for my taste.

File: src/fn_call.rs (hunk 1 above)
Comment (oli-obk) [reply src/fn_call.rs]: Having a few months perspective on this: Indeed "OMG WTF what was I thinking?" I'll teach the engine about dlsyms

Comment (RalfJung) [reply src/fn_call.rs]: I am implementing this now. (Latest `getrandom` needs it on macOS.)

**Comment by RalfJung** [comment]
@oli-obk Is there any point in keeping this open? I'd rather keep the PR list clean, having only things in there that are actually being worked on, or that are blocked on things that will likely soon be resolved.

**Pull request closed without merging**